8293067: (fs) Implement WatchService using system library (macOS)

mkartashev opened the pull request:
This is an implementation of `WatchService` based on File System Events API that is capable of generating events whenever a change occurs in an interesting directory or underneath it. Since the API naturally supports "recursive" watch, the `FILE_TREE` is supported by the watch service. 

Some things of note:
* There's one "service" thread per `WatchService` instance that is inactive unless changes occur in the watched directory. The changes are grouped by introducing a time delay between when they occurred and when they are reported, which is controlled by the sensitivity modifier of the watch service.
* Since FSEvents API reports directories only, the watch service keeps a snapshot (hierarchical if necessary) of the files in the directory being watched. The snapshot gets updated when an event in that directory or underneath it gets delivered. File changes are detected by comparing "last modified" time with a millisecond precision (`BasicFileAttributes.lastModifiedTime()`).
* There is a slight complication with the move of an entire directory hierarchy: FSEvents API only reports about the containing directory of that move and not about any of the directories actually moved. There's a separate test for that (`Move.java`).
* The code is careful not to do any I/O (such as reading the contents of a directory or attributes of a file) unless unavoidable. Any deviation from this line should be considered a bug (of, arguably, low priority).
* The native part consists mostly of API wrappers with one exception of the callback function invoked by the system to report the events that occurred. The sole task of the function is to convert from C strings to Java strings and pass the array of affected directories to Java code. This can be rewritten if desired to make the code more future-proof.

This commit leaves `PollingWatchService` unused. I'm not sure if I should/can do anything about it. Any advice is welcomed.

### Testing

* Tested by running `test/jdk/java/nio/file` and `test/jdk/jdk/nio` on MacOS 10.15.7 (x64) and `test/jdk/java/nio/` plus `test/jdk/jdk/nio` on MacOS 12.5.1 (aarch64).
* Also verified that new tests pass on Linux and Windows.
* This code (albeit in a slightly modified form) has been in use at JetBrains for around half a year and a few bugs have been found and fixed during that time period.

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [ ] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [ ] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [ ] Change requires CSR request [JDK-8296164](https://bugs.openjdk.org/browse/JDK-8296164) to be approved

### Issues
 * [JDK-8293067](https://bugs.openjdk.org/browse/JDK-8293067): (fs) Implement WatchService using system library (macOS)
 * [JDK-8296164](https://bugs.openjdk.org/browse/JDK-8296164): (fs) introduce system property to choose WatchService on macOS (**CSR**)


### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/10140/head:pull/10140` \
`$ git checkout pull/10140`

Update a local copy of the PR: \
`$ git checkout pull/10140` \
`$ git pull https://git.openjdk.org/jdk pull/10140/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 10140`

View PR using the GUI difftool: \
`$ git pr show -t 10140`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/10140.diff">https://git.openjdk.org/jdk/pull/10140.diff</a>

</details>

AlanBateman commented:
Thanks for this PR, I plan to review this but it will take some time. Just to set expectations that the patch will likely through through several iterations as there will be many issues to discuss.

AlanBateman commented:
> This commit leaves `PollingWatchService` unused. I'm not sure if I should/can do anything about it. Any advice is welcomed.

I think it is still used on on AIX but there is someone from IBM working on that. It may be that we will have to keep it for new ports. For macOS then we might have to introduce a system property to use it, at least until there is confidence with the new implementation.

mkartashev commented:
> For macOS then we might have to introduce a system property to use it, at least until there is confidence with the new implementation.

I actually had that property, but pulled it out before submitting the pull request, so it's going to be very easy to re-introduce it.

bplb commented:
In our CI system `test/jdk/java/nio/file/WatchService/Move.java` appears to time out consistently, although I have not seen it happen on my local machine.

mkartashev commented:
> In our CI system `test/jdk/java/nio/file/WatchService/Move.java` appears to time out consistently, although I have not seen it happen on my local machine.

Can you provide the `.jtr` file, please? I haven't seen the test hang, but by inserting different delays different systems can expose synchronization issues, of course.

bplb commented:
> Can you provide the `.jtr` file, please? I haven't seen the test hang, but by inserting different delays different systems can expose synchronization issues, of course.

I can't give the whole .jtr file due to internal content, but here are some excerpts.
```
/System/Volumes/Data/mesos/work_dir/jib-master/install/2022-09-07-1757010.brian.burkhalter.jdk/src.full/open/test/jdk/java/nio/file/WatchService/Move.java:39: warning: ExtendedWatchEventModifier is internal proprietary API and may be removed in a future release
import com.sun.nio.file.ExtendedWatchEventModifier;
                       ^
/System/Volumes/Data/mesos/work_dir/jib-master/install/2022-09-07-1757010.brian.burkhalter.jdk/src.full/open/test/jdk/java/nio/file/WatchService/Move.java:110: warning: ExtendedWatchEventModifier is internal proprietary API and may be removed in a future release
                    new WatchEvent.Kind<?>[]{ ENTRY_CREATE, ENTRY_DELETE },  ExtendedWatchEventModifier.FILE_TREE);
                                                                             ^
2 warnings
result: Passed. Compilation successful
```
```
execStatus=Error. Agent error\: java.lang.Exception\: Agent 5 timed out with a timeout of 480 seconds; check console log for any additional details
```
```
test result: Error. Agent error: java.lang.Exception: Agent 5 timed out with a timeout of 480 seconds; check console log for any additional details
```
The node in questions is a MacPro6_1 running Mac_OS_X_12.4 on a Intel_R__Xeon_R__CPU_E5-1620_v2___3.70GHz with total memory 17179869184 (16.00 GB).

Sadly, with timeouts, there is often little information to go on.

AlanBateman commented:
I did a first pass over this. It's a good start and demonstrates that the  file system events / FSEventStream* API can be used to implement WatchService.

It's really unfortunate that it requires calling CFRunLoopRun to do the run loop as that means an upcall and JNI code that we would normally try to keep out of this area. One concern is that there is a lot of UnixPath (bytes) <-> String <-> CFString conversions going on and these will need time to work through (even with UTF-8). Also 
CFRunLoopThread will need to be an InnocuousThread to avoid inheriting inheritable thread locals or context class loader.

One architectural issue is that the WatchService implementation should not be using Path.of or Files.* methods as that will cause a loop when interposing on the default provider. The provider implementation should only use use the Unix provider methods directly.

The use of toRealPath().normalize() looks odd, I would expect the normalize to be a no-op here.  This is an area that will require careful study as it's just not clear how the macOS work when the sym link is to another directory on another file system.

Look at ptr_to_jlong and jlong_to_jptr to go between points and jlong and that should eliminate some of the casts.

The code includes SUBTREE support. I had hope this wouldn't be included initially as it adds a bit more complexity and could be added later.

Style-wise the code is very different to the existing code and I think we'll need to do a few cleanup passes (overly long lines, naming, formatting, overdue of final, ... all minor stuff and there is a lot to go through before these things).

mkartashev commented:
@bplb Thanks for the log. It's really strange, though, that there is nothing from `System.out` because the test itself logs intensively even before any `WatchService` code is executed. Also, AFAIK timeouts trigger thread dump, which would be quite helpful. Are you not at liberty to share those as well?

mkartashev commented:
@AlanBateman 
Thank you for taking time to review this, much appreciated. I made some changes based on your feedback; see inlined comments below.

> It's really unfortunate that it requires calling CFRunLoopRun to do the run loop as that means an upcall and JNI code that we would normally try to keep out of this area. One concern is that there is a lot of UnixPath (bytes) <-> String <-> CFString conversions going on and these will need time to work through (even with UTF-8). Also CFRunLoopThread will need to be an InnocuousThread to avoid inheriting inheritable thread locals or context class loader.

Done, `CFRunLoopThread` is now `InnocuousThread`.

 
> One architectural issue is that the WatchService implementation should not be using Path.of or Files.* methods as that will cause a loop when interposing on the default provider. The provider implementation should only use use the Unix provider methods directly.

Duly noted and, I think, corrected.


> The use of toRealPath().normalize() looks odd, I would expect the normalize to be a no-op here. This is an area that will require careful study as it's just not clear how the macOS work when the sym link is to another directory on another file system.

I may be misinterpreting you here, but to me it seems `toRealPath()` is actually necessary. Let's say the watch root is a directory named "../symlink" that points to "/Users/maxim/work/tests/dir-watcher/". The FSEvents API will report strings denoting absolute path names like "/Users/maxim/work/tests/dir-watcher/subdir" and in order to be able to remove the common prefix (watch root), one has to make that watch root into an unambiguous absolute path name.

> Look at ptr_to_jlong and jlong_to_jptr to go between points and jlong and that should eliminate some of the casts.

All such cases were eliminated.

> The code includes SUBTREE support. I had hope this wouldn't be included initially as it adds a bit more complexity and could be added later.

The recursive subdirectory watch support was removed (for now); let's focus on the initial support for FSEvents.

> 
> Style-wise the code is very different to the existing code and I think we'll need to do a few cleanup passes (overly long lines, naming, formatting, overdue of final, ... all minor stuff and there is a lot to go through before these things).

 I also reformatted the native code and a bit of Java mostly to shorten the lines. The rest will have to be on case-by-case basis, I'm afraid.


This time around I only ran the `test/jdk/java/nio/file/WatchService` tests on MacOs 10.15.

bplb commented:
> @bplb Thanks for the log. It's really strange, though, that there is nothing from `System.out` because the test itself logs intensively even before any `WatchService` code is executed. Also, AFAIK timeouts trigger thread dump, which would be quite helpful. Are you not at liberty to share those as well?

@mkartashev When a test times out, often not a lot of information is captured, including `System.out` and thread dumps. It can be frustrating to debug.

bplb commented:
I reran the `java/nio/file/WatchService` tests with the current code and there were no failures this time. I wonder whether this might be due to the change from `CFRunLoopThread` to `InnocuousThread`? I had some similar timeouts with the kqueue-based prototype I developed which I think were due to problems with thread cleanup.

mkartashev commented:
> I reran the java/nio/file/WatchService tests with the current code and there were no failures this time. I wonder whether this might be due to the change from CFRunLoopThread to InnocuousThread?

Thanks! Let's hope it stays this way. 

BTW, my initial version did have a thread cleanup problem that was caught by other tests. That was because I assumed the run loop would exit as soon as all input sources are pulled from it; that assumption proved to be false, so now there's `MacOSXWatchService.runLoopStop()`.

bplb commented on src/java.base/macosx/classes/sun/nio/fs/MacOSXWatchService.java:
```diff
@@ -0,0 +1,634 @@
+     * Invoked on the CFRunLoopThread by the native code to report directories
+     * that need to be re-scanned.
+     */
+    private void callback(final long eventStreamRef,
```
Could this be named something more descriptive than `callback`, perhaps even `handleEvents` even though there is a method of the same name defined in `MacOSXWatchKey`?

mkartashev replied:
Sure.

bplb commented on src/java.base/macosx/classes/sun/nio/fs/MacOSXWatchService.java:
```diff
@@ -0,0 +1,634 @@
+
+                if (!relativeRootPath.equals(path)) {
+                    // Ignore events from subdirectories for now.
+                    continue;
```
Should a line
```
eventFlagsPtr += SIZEOF_FS_EVENT_STREAM_EVENT_FLAGS;
```
be added here? (The definition of `SIZEOF_FS_EVENT_STREAM_EVENT_FLAGS` would have to be moved up.)

mkartashev replied:
Yes, most certainly. Thanks for catching this!

bplb commented on src/java.base/macosx/classes/sun/nio/fs/MacOSXWatchService.java:
```diff
@@ -0,0 +1,634 @@
+                    continue;
+                }
+
+                final int flags = unsafe.getInt(eventFlagsPtr);
```
This looks like the only use of `Unsafe`. Could the objective be accomplished without using it, e.g., changing `long eventFlagsPtr` to a `long[]` or, maybe even simpler, a `boolean[]`?

mkartashev replied:
Earlier on the mailing list @AlanBateman wrote "It would help the review if the native code is just simple wrappers about the FSEvents as we'll likely replace a lot of the JNI code in this area in time." 
I understood that as a suggestion to keep the native code as simple as possible and, conversely, do as much as possible on the Java side. Reading native memory here was one of the decisions that were based on that understanding. 
I may have, of course, read too much into it, so I'm not opposed to implementing your suggestion. I would like, however, to have a confirmation that this direction of the changes is OK with @AlanBateman also.

bplb replied:
Your understanding is correct, but sometimes precisely mapping the native API can result in more complication than needed. While this might not be a strictly analogous example, consider the BSD system call `setattrlist(2)`. We only need to use it at present to set file times, so this call was exposed at the Java level with a simpler set of parameters. See [sun.nio.fs.BsdNativeDispatcher::setattrlist](https://github.com/openjdk/jdk/blob/5285035ed9bb43a40108e4d046e0de317730f193/src/java.base/macosx/classes/sun/nio/fs/BsdNativeDispatcher.java#L84
). I’ll leave it to @AlanBateman to express his own opinion on this.

bplb commented on src/java.base/macosx/native/libnio/fs/MacOSXWatchService.c:
```diff
@@ -0,0 +1,237 @@
+         __unused const FSEventStreamEventId eventIds[])
+{
+    JNIEnv *env = (JNIEnv *) JNU_GetEnv(jvm, JNI_VERSION_1_2);
+    if (!env) { // Shouldn't happen as run loop starts from Java code
```
Is this check, hence the `jvm` extern, really necessary?

mkartashev replied:
The check isn't strictly necessary, so I'm goign to drop it (there's a JNI version check in `nio_util.c`), but the `jvm` extern seems necessary because that's how `JNIEnv` is obtained and the latter is necessary to communicate the events back to the Java side.

bplb commented on src/java.base/macosx/native/libnio/fs/MacOSXWatchService.c:
```diff
@@ -0,0 +1,237 @@
+        // the thread-local reference to the watch service.
+        (*env)->CallVoidMethod(env, watchService, callbackMID,
+                               streamRef, javaEventPathsArray, eventFlags);
+    }
```
Is there a way to notify of events without an upcall? Maybe add the necessary values to a queue of some sort observed by the Java layer?

mkartashev replied:
I'm not sure I know how to synchronize native code with Java without any upcall to the Java side; but maybe you are suggesting something as simple as a `notify()` call in place of the current callback with many arguments?.

And also this goes against my understanding that the native part has to be kept as simple and as straightforward as possible. Having it  maintain a queue that has to be synchronized with Java will certainly complicate the native part.

bplb replied:
I was thinking of maybe something like a [BlockingQueue](https://docs.oracle.com/en/java/javase/19/docs/api/java.base/java/util/concurrent/BlockingQueue.html) field on `sun.nio.fs.MacOSXWatchService`, but I am not sure whether this is any more correct.

If the callback is retained, it might be worth considering using [JNU_CallMethodByName](https://github.com/openjdk/jdk/blob/5285035ed9bb43a40108e4d046e0de317730f193/src/java.base/share/native/libjava/jni_util.h#L134).

bplb commented on src/java.base/macosx/native/libnio/fs/MacOSXWatchService.c:
```diff
@@ -0,0 +1,237 @@
+    const FSEventStreamRef streamRef = jlong_to_ptr(eventStreamRef);
+
+    FSEventStreamStop(streamRef);       // Unregister with the FS Events service.
+                                        // No more callbacks from this stream.
```
Does `FSEventStreamUnscheduleFromRunLoop()` need to be called here?

mkartashev replied:
I think not. The [documentation](https://developer.apple.com/documentation/coreservices/1446990-fseventstreaminvalidate?language=objc) for `FSEventStreamInvalidate()` that gets called immediately after this line states the following:
> Invalidates the stream, like CFRunLoopSourceInvalidate() does for a CFRunLoopSourceRef. It will be unscheduled  from any runloops or dispatch queues upon which it had been scheduled.

So calling `FSEventStreamUnscheduleFromRunLoop()` seems superfluous; its semantics is described as
> ... removes the stream from the specified run loop, like CFRunLoopRemoveSource() does for a CFRunLoopSourceRef.

bplb replied:
That sounds correct.

turbanoff commented on src/java.base/macosx/classes/sun/nio/fs/MacOSXWatchService.java:
```diff
@@ -0,0 +1,636 @@
+
+        private static final long kFSEventStreamEventFlagRootChanged     = 0x00000020;
+
+        private final static Path relativeRootPath = theFS.getPath("");
```
let's use blessed modifiers order: `private static final`

mkartashev replied:
Sure! Fixed this one.

mkartashev commented:
@AlanBateman Do you have any follow-up comments? I'd like to move this one forward.

AlanBateman commented:
> @AlanBateman Do you have any follow-up comments? I'd like to move this one forward.

It's on my list to do another pass on this. I think the main thing that I'd like us to explore is using interfaces that would avoid the upcall. It would be good to see if there are any options that are feasible there.

mkartashev commented:
> It's on my list to do another pass on this. 

Thank you!

> I think the main thing that I'd like us to explore is using interfaces that would avoid the upcall. It would be good to see if there are any options that are feasible there.

It greatly depends on the definition of "upcall". I understand this as any call from native to Java code. The FSEvents API employs a callback-based approach, there's nothing to be explored there. 

However, as Brian suggested earlier, there are still options like reducing the complexity of the upcall to a mere `Object.notify()` or other simple means of synchronization. The side effect of such a change would be a much more complicated native portion of the callback, which I understood is something to be avoided. So I'm at a crossroads and would like a clearer guidance on the matter.

AlanBateman commented:
>  The FSEvents API employs a callback-based approach, there's nothing to be explored there.

The suggestion was to explore doing this at a low level with kqueue so that we don't have callbacks.  I think this was the direction that Brian was exploring at one point.

I do plan to do a detailed walk through of the FSEvents prototype.

mkartashev commented:
> The suggestion was to explore doing this at a low level with kqueue so that we don't have callbacks.

Vladimir Kempik had a draft implementation based on kqueue that, AFAIR, passed all the existing tests; he may be interested in getting his work merged into the mainline.

My primary concern with this approach is that it doesn't give a lot of advantage over `PollingWatchService` when you need to watch for changes in a very deep hierarchy of directories, the main use case for my employer. At the same time FSEvents API naturally provide the ability for recursive watch almost for free resource-wise.

AlanBateman commented:
> My primary concern with this approach is that it doesn't give a lot of advantage over `PollingWatchService` when you need to watch for changes in a very deep hierarchy of directories, the main use case for my employer. At the same time FSEvents API naturally provide the ability for recursive watch almost for free resource-wise.

Do you know if FSEvents is implemented on a supported/document interface? I think it would be useful to know if there are  alternatives that provide a way to block for events so we don't have the callback.

VladimirKempik commented:
kqueue implementation is generally bad idea, example of implementation: https://github.com/VladimirKempik/jdk/commit/507f8acecb2d48d0673d63386b1959ddb8c040a3
, first of all:
1) The kqueue monitor requires a file descriptor to be opened for every file being watched. As a result, this monitor scales badly with the number of files being observed and may begin to misbehave as soon as the process runs out of file descriptors. the limit is about 10k files

2) this modifier StandardWatchEventKinds.ENTRY_MODIFY says to monitor changes of individual files in the dir, but kevent on the dir only reports changes to the dir itself ( e.g. file added, file removed) and won't says anything about changes to file in the dir ( file modified). As a result we still have to use good old periodic polling on the directories and can miss some event ( e.g. file create and removed immediately)

mkartashev commented:
> Do you know if FSEvents is implemented on a supported/documenedt interface? 

Unfortunately, my knowledge doesn't go deep into OSX. FWIW, I haven't seen anything like that when I was researching for this implementation.

bplb commented:
> My primary concern with this approach is that it doesn't give a lot of advantage over `PollingWatchService` [...].

Has there been any form of bechmarking comparison other than simple subjective observation?

mkartashev commented:
> Has there been any form of bechmarking comparison other than simple subjective observation?

Yes, but it was about 9 months ago and I didn't save the data. Since Vladimir kept the [kqueues-based implementation](https://github.com/VladimirKempik/jdk/commit/507f8acecb2d48d0673d63386b1959ddb8c040a3), anyone willing to experiment and collect new data can do so relatively easily.

VladimirKempik commented:
>but modifying a contained file alone will not generate any event on the directory descriptor, hence watching for `ENTRY_MODIFY` requires an open descriptor for each file in the directory.

exactly, the main reason I have dropped that work.

bplb commented:
@VladimirKempik Thank you for corroborating my findings.

bplb commented:
> My primary concern with this approach is that it doesn't give a lot of advantage over `PollingWatchService` when you need to watch for changes in a very deep hierarchy of directories

Could the upcall be replaced for example with a `socketpair` where the native code writes the event data to one end, and the Java code polls the other end and reads the event data?

mkartashev commented:
> Could the upcall be replaced for example with a socketpair 

Yes, this should be possible; I think I saw something of that sort implemented in the Linux watch service.

bplb commented:
On Oct 5, 2022, at 11:42 PM, Maxim Kartashёv ***@***.******@***.***>> wrote:

Could the upcall be replaced for example with a socketpair

Yes, this should be possible; I think I saw something of that sort implemented in the Linux watch service.

Yes, it simultaneously polls the inotify descriptor and one end of a socketpair, the latter being used for wakeup.

AlanBateman commented:
It's a real shame that it requires use of a CFRunLoop and callbacks but if that is the only facility that it provides then I suppose we have to use it. I see the discussion about using a socketpair but I don't think that is sensible thing to do here. The socketpair that is used elsewhere is for wakeup only, similar to what we do in the Selector implementations. So I think we should move forward with the current approach but probably will need a few rounds of cleanup.

mkartashev commented on test/jdk/jdk/nio/zipfs/test.policy:
```diff
@@ -4,4 +4,5 @@ grant {
     permission java.util.PropertyPermission "test.src","read";
     permission java.util.PropertyPermission "user.dir","read";
     permission java.lang.RuntimePermission "accessUserInformation";
+    permission java.lang.RuntimePermission "loadLibrary.nio";
```
I understand that previously using a `WatchService` didn't require a native library to be loaded, but now it does.

mkartashev commented on src/java.base/share/native/libnio/nio_util.c:
```diff
@@ -26,11 +26,15 @@
 DEF_JNI_OnLoad(JavaVM *vm, void *reserved)
 {
     JNIEnv *env;
+    jvm = vm;
```
Of course. Fixed.

turbanoff commented on test/jdk/java/nio/file/WatchService/Move.java:
```diff
@@ -0,0 +1,208 @@
+        Path dir = TestUtil.createTemporaryDirectory();
+        try {
+            testMoveSubtree(dir);
+        } catch(UnsupportedOperationException e) {
```
```suggestion
        } catch (UnsupportedOperationException e) {
```

turbanoff commented on test/jdk/java/nio/file/WatchService/Move.java:
```diff
@@ -0,0 +1,208 @@
+        dir = TestUtil.createTemporaryDirectory();
+        try {
+            testMoveFileToDirectory(dir);
+        } catch(UnsupportedOperationException e) {
```
```suggestion
        } catch (UnsupportedOperationException e) {
```

turbanoff commented on src/java.base/macosx/classes/sun/nio/fs/MacOSXWatchService.java:
```diff
@@ -0,0 +1,636 @@
+            directorySnapshot = DirectorySnapshot.create(getRealRootPath());
+
+            synchronized (eventStreamRefLock) {
+                final int kFSEventStreamCreateFlagWatchRoot  = 0x00000004;
```
```suggestion
                final int kFSEventStreamCreateFlagWatchRoot = 0x00000004;
```

mkartashev replied:
Thanks! I'll wait for more formatting-related input from @AlanBateman and others and then do all the changes in bulk.

bplb commented on src/java.base/macosx/classes/sun/nio/fs/BsdFileSystem.java:
```diff
@@ -54,8 +52,7 @@ class BsdFileSystem extends UnixFileSystem {
     {
-        // use polling implementation until we implement a BSD/kqueue one
-        return new PollingWatchService();
+        return new MacOSXWatchService();
```
Is it still intended that there be a system property to enable `MacOSXWatchService` which would be off by default?

mkartashev replied:
There was such a suggestion (not sure about "off by default", though). I guess this should go through CSR review, in which case I'm going to need some guidance.

bplb replied:
If you file a CSR then I can help with editing.

mkartashev replied:
Filed https://bugs.openjdk.org/browse/JDK-8296164
@bplb Could you please take a look and edit parts that need it?

bplb replied:
@mkartashev I made some minor changes to the CSR. Was the property name previously agreed upon? I would tend to drop the `.use` from the end of it.

mkartashev replied:
> Was the property name previously agreed upon?

Nope. Please, rename as you see fit, I really have no preference nor do I have any experience in CSRs.

AlanBateman replied:
If we can get the new implementation stable then it might not be worth including the polling implementation in the macOS build. The make file that controls whether PollingWatchService is included or not is in make/modules/java.base/Java.gmk.

bplb commented on src/java.base/macosx/classes/sun/nio/fs/MacOSXWatchService.java:
```diff
@@ -0,0 +1,636 @@
+
+class MacOSXWatchService extends AbstractWatchService {
+    private static final MacOSXFileSystemProvider theFSProvider = DefaultFileSystemProvider.instance();
+    private static final MacOSXFileSystem theFS = (MacOSXFileSystem) theFSProvider.theFileSystem();
```
Typically `static final` constants are presented in upper case with underscores as for example `THE_FS_PROVIDER`.

bplb commented on src/java.base/macosx/classes/sun/nio/fs/MacOSXWatchService.java:
```diff
@@ -0,0 +1,636 @@
+        t.setDaemon(true);
+        t.start();
+
+        // In order to be able to schedule any FSEventStream's, a reference to a run loop is required.
```
Nit: I think it should be "FSEventStreams", i.e., no apostrophe.

bplb commented on src/java.base/macosx/classes/sun/nio/fs/MacOSXWatchService.java:
```diff
@@ -0,0 +1,636 @@
+
+        private static final long kFSEventStreamEventFlagRootChanged     = 0x00000020;
+
+        private static final Path relativeRootPath = theFS.getPath("");
```
Same comment as for lines 51-52. Unsure of line 311 however as that refers to a specific constant. Should that be perhaps in `UnixConstants`?

mkartashev replied:
I'll change the spelling of all but `kFSEventStreamEventFlagRootChanged`, which I'd rather keep as is. The constant being spelled exactly the same as in the corresponding C header file makes it easier to check its correctness (https://developer.apple.com/documentation/coreservices/1455361-fseventstreameventflags/kfseventstreameventflagrootchanged?language=objc).
As for moving it to `UnixConstants`, I'm also not so sure as it is hardly a "Unix" constant, more like a macOS constant.

mkartashev replied:
How about I make `kFSEventStreamEventFlagRootChanged` into a local variable? It's only used in one place and doesn't have to be a static field really.

bplb replied:
Sounds good.

bplb commented on src/java.base/macosx/classes/sun/nio/fs/MacOSXWatchService.java:
```diff
@@ -0,0 +1,636 @@
+        }
+
+        private Path toRelativePath(final String absPath) {
+            return   (absPath.length() > realRootPathLength)
```
Nit: Are there several spaces after `return`?

mkartashev replied:
Yes, to align the beginning of the returned expression with its continuations on the subsequent two lines.

bplb replied:
OK

mkartashev commented:
@AlanBateman Are there any outstanding issues that prevents this from being merged, apart from maybe the implementation of the [CSR](https://bugs.openjdk.org/browse/JDK-8296164)? I believe I addressed all the concerns raised in your comment from Sep, 11, but if more have been found since then, please do let me know.

AlanBateman commented:
> @AlanBateman Are there any outstanding issues that prevents this from being merged, apart from maybe the implementation of the [CSR](https://bugs.openjdk.org/browse/JDK-8296164)? I believe I addressed all the concerns raised in your comment from Sep, 11, but if more have been found since then, please do let me know.

I've been tanked with other items. This PR is on my list to review. I would like to see this change in JDK 20.

VladimirKempik commented:
> _Mailing list message from [Michael Hall](mailto:mik3hall@gmail.com) on [nio-dev](mailto:nio-dev@mail.openjdk.org):_
> 
> I had a thought for some code to try this on but seem to get a number of build errors now.
> 
> http://mikehall.pairserver.com/errors10140.txt <http://mikehall.pairserver.com/errors10140.txt>
> 
> Using the GitHub Desktop app to change to pr/10140
> 
> -------------- next part -------------- An HTML attachment was scrubbed... URL: <https://mail.openjdk.org/pipermail/nio-dev/attachments/20221102/d5ad452a/attachment.htm>

You are using xcode 14 with sdk for macos Ventura hence having issues. that's orthogonal to this PR.

AlanBateman commented:
I don't test/jdk/java/nio/file/WatchService/JNIChecks.java should be integrated as is. It's okay to run tests with -Xcheck:jni but it would be a lot more maintainable to just add `@run main/othervm -Xcheck:jni XXX` to the existing tests.  Basic.java is the main unit test for this API and having that re-run with -Xcheck:jni may be enough here. It's relatively long running on some platforms so we could put give its its own run tag, as in:

```
-/* @test
+/* @test id = default
  * @bug 4313887 6838333 7017446 8011537 8042470
  * @summary Unit test for java.nio.file.WatchService
  * @library ..
  * @run main Basic
  */
 
+/* @test id=checkjni
+ * @library ..
+ * @run main/other -Xcheck:jni Basic
+ */
+
```

Note that this runs it with -Xcheck:jni on all platforms, it doesn't need to be restricted to macOS.

mkartashev commented:
> Basic.java is the main unit test for this API and having that re-run with -Xcheck:jni may be enough here. 
I'm OK with running just `Basic.java`, but I don't understand how adding `-Xcheck:jni` to `@run` will make this work because we need the test to check the output for warnings from those checks.

AlanBateman commented:
> In our CI system `test/jdk/java/nio/file/WatchService/Move.java` appears to time out consistently, although I have not seen it happen on my local machine.

I've been backed up with other things and finally getting to look at this more closely. For testing, I tried both release and debug builds across a range of macOS releases on both x64 and aarch64. Unfortunately there is a lot of timeouts and intermittent failures and across quite a range of macOS releases (from 10.15 to 12.2).

The most frequent failure seems to be the testOverflowEvent test in LotsOfEvents.  Here's an example failure:
```
----------System.out:(1/47)----------
Seed from RandomFactory = 1932318894124342221L
----------System.err:(12/655)----------
java.lang.RuntimeException: Key not signalled (unexpected)
	at LotsOfEvents.drainAndCheckOverflowEvents(LotsOfEvents.java:105)
	at LotsOfEvents.testOverflowEvent(LotsOfEvents.java:78)
	at LotsOfEvents.main(LotsOfEvents.java:51)
	at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:104)
	at java.base/java.lang.reflect.Method.invoke(Method.java:578)
	at com.sun.javatest.regtest.agent.MainActionHelper$AgentVMRunnable.run(MainActionHelper.java:312)
	at java.base/java.lang.Thread.run(Thread.java:1591)
```

mkartashev commented:
@AlanBateman still can't reproduce any of that on my hardware, so I added some tracing so that next time tests are run we can at least see what's happening.

AlanBateman commented:
> @AlanBateman still can't reproduce any of that on my hardware, so I added some tracing so that next time tests are run we can at least see what's happening.

LotsOfEvents.java on macOS 12.2
```
----------stdout:(7/337)----------
WatchService: get run loop -> 600003a5c500
WatchService: running run loop 600003a5c500
WatchService: created event stream 101105150
WatchService: scheduled event stream 101105150 on run loop 600003a5c500
WatchService: running run loop 600003a5c500
WatchService: stopped event stream 101105150
WatchService: stopped run loop 600003a5c500
----------System.out:(10/421)----------
Seed from RandomFactory = 5027502271796047304L
WatchService: Run loop 600003a5c500 - waiting...
WatchService: Run loop 600003a5c500 - starting...
Test: dir.register - start
WatchService: Run loop 600003a5c500 - waiting for event source...
Test: dir.register - done
WatchService: Run loop 600003a5c500 - starting...
Test: created files
Test: About to poll
WatchService: Run loop 600003a5c500 - waiting for event source...
----------System.err:(12/655)----------
java.lang.RuntimeException: Key not signalled (unexpected)
	at LotsOfEvents.drainAndCheckOverflowEvents(LotsOfEvents.java:112)
	at LotsOfEvents.testOverflowEvent(LotsOfEvents.java:84)
	at LotsOfEvents.main(LotsOfEvents.java:51)
	at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:104)
	at java.base/java.lang.reflect.Method.invoke(Method.java:578)
	at com.sun.javatest.regtest.agent.MainActionHelper$AgentVMRunnable.run(MainActionHelper.java:312)
	at java.base/java.lang.Thread.run(Thread.java:1591)

JavaTest Message: Test threw exception: java.lang.RuntimeException
JavaTest Message: shutting down test

result: Failed. Execution failed: `main' threw exception: java.lang.RuntimeException: Key not signalled (unexpected)
```

DeleteInterference.java also  on macOS 12.2, this one is a timeout:
```
open 292 end
open 293 begin
WatchService: Run loop 6000004a9000 - waiting...
WatchService: Run loop 6000004a9000 - starting...
WatchService: Run loop 6000004a9000 - waiting for event source...
open 293 end
open 294 begin
WatchService: Run loop 6000004ade00 - waiting...
WatchService: Run loop 6000004ade00 - starting...
WatchService: Run loop 6000004ade00 - waiting for event source...
WatchService: Run loop 6000004ade00 - starting...
WatchService: Run loop 60000
result: Error. Agent error: java.lang.Exception: Agent 8 timed out with a timeout of 480 seconds; check console log for any additional details

test result: Error. Agent error: java.lang.Exception: Agent 8 timed out with a timeout of 480 seconds; check console log for any additional details
```

The .jtr are huge due to all the tracing.

mkartashev commented:
@AlanBateman Thank you for the log. I see all the right calls, but I don't see any events coming in from the OS.

I'm afraid this is the line beyond which I can't tread; there seem to be instances where [certain directories don't generate FSEvents](https://github.com/livereload/livereload-site/blob/master/livereload.com/_articles/troubleshooting/os-x-fsevents-bug-may-prevent-monitoring-of-certain-folders.md), there's even a [tool](https://github.com/andreyvit/find-fsevents-bugs) to find those. Unfortunately, I know nothing beyond what was said on those pages.

jaikiran commented:
> @AlanBateman Thank you for the log. I see all the right calls, but I don't see any events coming in from the OS.
> 
> I'm afraid this is the line beyond which I can't tread; there seem to be instances where [certain directories don't generate FSEvents](https://github.com/livereload/livereload-site/blob/master/livereload.com/_articles/troubleshooting/os-x-fsevents-bug-may-prevent-monitoring-of-certain-folders.md), there's even a [tool](https://github.com/andreyvit/find-fsevents-bugs) to find those. Unfortunately, I know nothing beyond what was said on those pages.

FWIW - I am seeing the same exceptions that Alan mentioned across various WatchService related tests on different versions of macos, with the changes in this PR. That isn't too surprising because I use the same infrastructure as Alan to test this. Looking at some of the debug logs that I have access too, it does appear that the OS itself isn't generating those events (or is generating it in a very delayed manner). I see that in your changes, you use a latency of `0.5` seconds, but the delay I'm seeing in the logs is far greater than that. I'm not yet sure what's going on.
While reviewing the native APIs being used in this change, I noticed that `FSEventStreamScheduleWithRunLoop` is called in the implementation. The macOS documentation for this API states that this is deprecated (since macOS 10.5) https://developer.apple.com/documentation/coreservices/1447824-fseventstreamschedulewithrunloop. Should we instead be implementing this using some other APIs that macOS recommends?

mkartashev commented:
I just noticed a curious line in the documentation for `FSEventStreamStart()`:
>  It ought to always succeed, but in the event it does not then your code should fall back to performing recursive scans of the directories of interest as appropriate.

So I added a check for the return value of `FSEventStreamStart()`; if anyone who has seen tests failing (@AlanBateman @jaikiran) could please re-run some of them to see if an `InternalError` gets thrown, we can at least establish that the fallback mechanism the documentation talks about actually has to be implemented, which will give us a way to move forward.

Thanks in advance!

jaikiran commented:
> I just noticed a curious line in the documentation for `FSEventStreamStart()`:
> 
> > It ought to always succeed, but in the event it does not then your code should fall back to performing recursive scans of the directories of interest as appropriate.
> 
> So I added a check for the return value of `FSEventStreamStart()`; if anyone who has seen tests failing (@AlanBateman @jaikiran) could please re-run some of them to see if an `InternalError` gets thrown, we can at least establish that the fallback mechanism the documentation talks about actually has to be implemented, which will give us a way to move forward.
> 
> Thanks in advance!

I gave this latest state of the PR a try and the tests continue to fail across various versions of the macos, like previously. I even noticed a JVM crash:

```
#
# A fatal error has been detected by the Java Runtime Environment:
#
#  SIGSEGV (0xb) at pc=0x00007fff206a1b2e, pid=xxxx, tid=xxxx
#
# JRE version: Java(TM) SE Runtime Environment (20.0) (build 20-internal-2022-11-29)
# Java VM: Java HotSpot(TM) 64-Bit Server VM (20-internal-2022-11-29, mixed mode, sharing, tiered, compressed oops, compressed class ptrs, g1 gc, bsd-amd64)
# Problematic frame:
C [CoreFoundation+0x14ab2e]  _CFGetNonObjCTypeID+0xa
```
(it could be that the JVM crash was present even in previous version of this PR before your latest commit - I hadn't looked into all log files).

mkartashev commented:
@jaikiran The tests were supposed to continue to fail, but this time there was a chance to diagnose the failure by checking if an InternalError was thrown from my latest commit. Have you seen any in the test logs?

jaikiran commented:
> @jaikiran The tests were supposed to continue to fail, but this time there was a chance to diagnose the failure by checking if an InternalError was thrown from my latest commit. Have you seen any in the test logs?

What I meant in my earlier comment was the tests failed with the same errors/timeouts as previously. There was no InternalError thrown.

mkartashev commented:
> What I meant in my earlier comment was the tests failed with the same errors/timeouts as previously. There was no InternalError thrown.

Thanks! No luck, then...

mkartashev commented:
> Would using the non-deprecated API be something worth investigating?

That's on my to-do list.

> In the meantime, I will try and find some more time to see if we can narrow down these failures on the systems where this is failing. I don't know how soon I can get to that.

Thanks anyway, much appreciated!

mkartashev commented:
> Is there any chance this will be made available in some other way?
[JetBrains Runtime](https://github.com/JetBrains/JetBrainsRuntime) has FSEvents-based implementation of WatchService on macOS.

AlanBateman commented:
> [JetBrains Runtime](https://github.com/JetBrains/JetBrainsRuntime) has FSEvents-based implementation of WatchService on macOS.

Is this similar to what was proposed here, I'm curious if the same issues were encountered when running the tests across a wider range of macOS releases.

mkartashev commented:
@AlanBateman It is the implementation this one was based upon. No _known_ bugs exist there at this moment in time (tests were run on IntelliJ infrastructure, of course).

bplb commented:
@mkartashev Do you intend to continue work on this PR? If not, then we could potentially take over the work if that is acceptable.

bplb commented:
> It lacks a WatchService ,my own attempt at a kqueue one for that coming up short.

As was previously discussed (I think) I don't believe that a `kqueue` implementation is feasible as it will eventually run out of file descriptors. This is specifically due to `MODIFY` which cannot be detected without having an open file descriptor for each monitored file.

mkartashev commented:
> @mkartashev Do you intend to continue work on this PR? If not, then we could potentially take over the work if that is acceptable.

I plan to have a shot at using newer (not declared deprecated) API for scheduling the event stream. If that fails, this implementation is up for grabs for all I care.

mkartashev commented:
> How will your changes be tested if the PR has been closed?

I guess I'll re-open it, perhaps as a new one if necessary. As you correctly noted, I can't test this fully by myself so I have to make the code available to others somehow.

bplb commented:
> [...] I can't test this fully by myself so I have to make the code available to others somehow.

I think that we can help with testing.

mkartashev commented:
For the record: I finally got around to trying the new dispatch queues API. This reduces the complexity of the current implementation quite a bit and passes all tests except one. The obstacle I have not been able to overcome is this: the new implementation consistently fails the test `test/jdk/java/nio/file/WatchService/LotsOfCancels.java`, which creates many watch services and constantly starts/stops FSEventStream's.

I have not been able to find any solution to this; perhaps we are hitting some OS resource limit, even though not that many streams are active at any given moment (sometimes you can also see the "too many open files" exception from that test when it creates new temporary files).

So far it loos like this new API is no go.
